gitlab/.gitlab-ci.yml

```diff
@@ -0,0 +1,417 @@
+          mkdir -p deploy/iso;
+          cp -f ${BUILD_DIR}/deploy/*.iso deploy/iso;
+        fi
+    # Preserve the contents of the deploy directory
+    - cp -Rf ${BUILD_DIR}/deploy/* deploy/
+    - if [[ "x$DEBUG" != "x" ]]; then ls -l deploy/*; fi
+  dependencies:
+    - setup:workspace
+    - build:dom0:image
+    - build:upgrade:image
+    - build:uivm:image
+    - build:ndvm:image
+    - build:syncvm:image
+  artifacts:
+    paths:
+      - deploy/iso
+    when: always
+    expire_in: 2 days
+
+stage:iso:
+  <<: *stage_definition
+  variables:
+    IMAGE_TO_BUILD: "iso"
+  only:
+    variables:
+      - $PACKAGE_TYPE == "iso"
+
+stage:pxe:
+  <<: *stage_definition
+  variables:
+    IMAGE_TO_BUILD: "pxe"
+  only:
+    variables:
+      - $PACKAGE_TYPE == "pxe"
+
+stage:all:
+  <<: *stage_definition
+  variables:
+    IMAGE_TO_BUILD: "iso pxe repository"
+  before_script: *setup_cm
+  dependencies:
+    - setup:workspace
+    - build:all
+  only:
+    variables:
+      - $PACKAGE_TYPE == "all"
+
+.deploy_template: &deploy_definition  # Template for saving the iso to S3
+  image: alpine:latest
+  stage: deploy
+  before_script:
+    # Install the coreutils package and awscli python package, this enables easy upload to S3
+    - apk -q update && apk -qUu add py-pip && pip install --quiet awscli
+  script:
+    - if [[ "x$DEBUG" != "x" ]]; then echo "ls -al deploy"; ls -lR deploy/; fi
+    # Upload the iso to a directory named after today and the branch. Make it expire in $NUM_DAYS_TO_EXPIRE days, defaults to 30
+    - |
+      if [ -d deploy/iso ]; then
+        let EXPIRE_HOURS=$NUM_DAYS_TO_EXPIRE*24;
+        EXPIRATION_DATE=`date -d "$EXPIRE_HOURS:00:00" -I`;
+        TODAY=`date -I`;
+        aws s3 cp deploy/iso/* s3://${S3_BUCKET}/iso/$TODAY/$CI_COMMIT_REF_NAME/ --sse-kms-key-id $AWS_S3_ISO_KEY_ID --sse aws:kms --expires $EXPIRATION_DATE;
+      fi
+
+deploy:iso:
+  <<: *deploy_definition
+  dependencies:
+    - setup:workspace
+    - stage:iso
+  only:
+    variables:
+      - $PACKAGE_TYPE == "iso"
+
+deploy:all:
+  <<: *deploy_definition
+  dependencies:
+    - setup:workspace
+    - stage:all
+  only:
+    variables:
+      - $PACKAGE_TYPE == "all"
+
+cleanup:sstate:kernel:
+  image: debian:latest
+  stage: prepare
+  script:
+    # Delete anything that has not been accessed in 4 days
+    - find /cache/sstate/${CI_COMMIT_REF_NAME} -name "*kernel*" -delete
+    - find /cache/sstate/${CI_COMMIT_REF_NAME} -name "*argo-module*" -delete
+    - find /cache/sstate/${CI_COMMIT_REF_NAME} -name "*txt-info-module*" -delete
+    - find /cache/sstate/${CI_COMMIT_REF_NAME} -name "*linux-openxt*" -delete
+  when: manual
+
+cleanup:sstate:prune:
+  image: debian:latest
+  stage: prepare
+  script:
+    # Delete anything that has not been accessed in 4 days
+    - find /cache/sstate/${CI_COMMIT_REF_NAME} -type f -atime +4 -delete
+    - find /cache/sstate/${CI_COMMIT_REF_NAME} -type d -atime +4 -exec rm -rf {} \;
```

Comment:
This is a little confusing to me though:
- Remove all files that have not been modified in the last elapsed consecutive 4 days.
- Remove all the directories (including content) that had no file added, removed or renamed in the last elapsed consecutive 4 days.

So the second find command might delete files that were not deleted by the first rule. Yet that might not be happening due to the `ipk` naming and versioning convention.

Reply:
I admit I haven't tested this one, my sstate doesn't usually last long enough to test this. I think the intention was to just remove the directory if it was emptied from the first `find` command.

Reply:
Right, that makes sense, then `find /cache/sstate/${CI_COMMIT_REF_NAME} -type d -empty -delete` would take care of that.
I do not think this problem happens due to the ipk filenames including the version though.